lalegpl/lib/lale/nsga2.py

```diff
@@ -0,0 +1,567 @@
+                fitnessValue = [np.inf for i in range(nscorers)]
+                logger.info(f"Train fitnessValue - {fitnessValue}")
+
+            return fitnessValue
+
+        def time_check_callback(alg):
+            current_time = time.time()
+            elapsed_time = current_time - opt_start_time
+            logger.info(f"NFE Complete - {alg.nfe}, Elapsed Time - {elapsed_time}")
+
+        parameter_num = len(self.model_helper.param_choices)
+        target_num = len(self.scoring)
+        # Adjust max_evals if not a multiple of population size. This is
+        # required as Platypus performs evaluations in multiples of
+        # population_size.
+        adjusted_max_evals = (
+            self.max_evals // self.population_size
+        ) * self.population_size
+        if adjusted_max_evals != self.max_evals:
+            logger.info(
+                f"Adjusting max_evals to {adjusted_max_evals} from specified {self.max_evals}"
+            )
+
+        problem = Problem(parameter_num, target_num)
+        problem.types[:] = self.model_helper.types
+        problem.function = train_test_model
+
+        # Set the variator based on types of decision variables
+        varg = {}
+        first_type = problem.types[0].__class__
+        all_type_same = all([isinstance(t, first_type) for t in problem.types])
+        # use compound operator for mixed types
+        if not all_type_same:
+            varg["variator"] = CompoundOperator(SBX(), HUX(), PM(), BitFlip())
+
+        algorithm = NSGAII(
+            problem,
+            population_size=self.population_size,
+            **varg,
+        )
+
+        try:
+            algorithm.run(adjusted_max_evals, callback=time_check_callback)
+        except MaxBudgetExceededException as e:
+            logger.warning(
+                f"Max optimization time budget exceeded. Optimization exited prematurely.\n{e}"
+            )
+
+        solutions = nondominated(algorithm.result)
+        # solutions = [s for s in algorithm.result if s.feasible]`
+        # solutions = algorithm.result
+
+        moo_solutions = []
+        for solution in solutions:
+            vars = []
+            for pnum in range(parameter_num):
+                vars.append(problem.types[pnum].decode(solution.variables[pnum]))
+
+            vars_dict = self.param_to_dict(
+                vars,
+                self.model_helper.param_choices,
+                self.model_helper.param_categories,
+                self.model_helper.param_type,
+            )
+            moo_solutions.append(self.Soln(vars_dict, solution.objectives))
+            logger.info(f"{vars}, {solution.objectives}")
+
+        self.moo_solutions = moo_solutions
+
+        pareto_models = []
+        for solution in self.moo_solutions:
+            est = self.model_helper.create_instance(solution.variables)
+            est_trained = est.fit(X, y)
+            pareto_models.append((solution.variables, est_trained))
```

Comment:
There might be many many pareto models and training all of them might take up a lot of time post-optimization. It might be worthwhile to limit this number (for example, 10 or 25). Then we can somehow pick at most that many pareto solutions and train them. Picking the pareto solutions might be some additional work (but much cheaper computationally relative to the actual MOO).